File: tests/python/test_gatorosc.py
"""
Tests for the GatorOsc indicator Python bindings.
"""

import pytest
import numpy as np
import my_project as ta

def test_gatorosc_basic():
    """Test basic GatorOsc calculation."""
    # Create sample data
    data = np.random.randn(100) * 10 + 50
    
    # Calculate GatorOsc
    upper, lower, upper_change, lower_change = ta.gatorosc(data)
    
    # Check output shape
    assert len(upper) == len(data)
    assert len(lower) == len(data)
    assert len(upper_change) == len(data)
    assert len(lower_change) == len(data)
    
    # Check that we have some non-NaN values after warmup
    assert not np.all(np.isnan(upper))
    assert not np.all(np.isnan(lower))
    assert not np.all(np.isnan(upper_change))
    assert not np.all(np.isnan(lower_change))
    
    # Check that upper is positive and lower is negative
    valid_upper = upper[~np.isnan(upper)]
    valid_lower = lower[~np.isnan(lower)]
    assert np.all(valid_upper >= 0)
    assert np.all(valid_lower <= 0)

def test_gatorosc_custom_params():
    """Test GatorOsc with custom parameters."""
    data = np.random.randn(200) * 10 + 50
    
    # Test with custom parameters
    upper, lower, upper_change, lower_change = ta.gatorosc(
        data, 
        jaws_length=21, 
        jaws_shift=10,
        teeth_length=13,
        teeth_shift=7,
        lips_length=8,
        lips_shift=5
    )
    
    assert len(upper) == len(data)
    assert len(lower) == len(data)
    assert len(upper_change) == len(data)
    assert len(lower_change) == len(data)

def test_gatorosc_errors():
    """Test GatorOsc error handling."""
    # Empty data
    with pytest.raises(ValueError):
        ta.gatorosc(np.array([]))
    
    # All NaN data
    with pytest.raises(ValueError):
        ta.gatorosc(np.full(50, np.nan))
    
    # Invalid parameters
    with pytest.raises(ValueError):
        ta.gatorosc(np.random.randn(50), jaws_length=0)

def test_gatorosc_stream():
    """Test GatorOsc streaming functionality."""
    stream = ta.GatorOscStream()
    
    # Initial updates should return None
    # With default params: jaws_length=13, jaws_shift=8, teeth_length=8, teeth_shift=5, lips_length=5, lips_shift=3
    # lower_change_warmup = 13, so first output at index 12
    for i in range(20):
        result = stream.update(50.0 + i)
        if i < 12:  # Before minimum warmup (lower_change_warmup = 13, first output at index 12)
            assert result is None
    
    # Eventually should get results
    for i in range(50):
        result = stream.update(50.0 + i)
        if result is not None:
            upper, lower, upper_change, lower_change = result
            assert isinstance(upper, float)
            assert isinstance(lower, float)
            assert isinstance(upper_change, float)
            assert isinstance(lower_change, float)
            assert upper >= 0
            assert lower <= 0

def test_gatorosc_stream_custom_params():
    """Test GatorOsc streaming with custom parameters."""
    stream = ta.GatorOscStream(
        jaws_length=21,
        jaws_shift=10,
        teeth_length=13,
        teeth_shift=7,
        lips_length=8,
        lips_shift=5
    )
    
    # Feed some data
    for i in range(50):
        result = stream.update(50.0 + i * 0.1)
        # Just verify it doesn't crash and eventually returns values

def test_gatorosc_batch():
    """Test GatorOsc batch calculation."""
    data = np.random.randn(100) * 10 + 50
    
    # Test batch with multiple parameter combinations
    result = ta.gatorosc_batch(
        data,
        jaws_length_range=(10, 15, 5),
        jaws_shift_range=(6, 10, 2),
        teeth_length_range=(6, 10, 2),
        teeth_shift_range=(3, 6, 3),
        lips_length_range=(3, 6, 3),
        lips_shift_range=(2, 4, 2)
    )
    
    # Check result structure
    assert 'upper' in result
    assert 'lower' in result
    assert 'upper_change' in result
    assert 'lower_change' in result
    assert 'jaws_lengths' in result
    assert 'jaws_shifts' in result
    assert 'teeth_lengths' in result
    assert 'teeth_shifts' in result
    assert 'lips_lengths' in result
    assert 'lips_shifts' in result
    
    # Check shapes
    n_combos = 2 * 3 * 3 * 2 * 2 * 2  # Product of range sizes
    assert result['upper'].shape == (n_combos, len(data))
    assert result['lower'].shape == (n_combos, len(data))
    assert result['upper_change'].shape == (n_combos, len(data))
    assert result['lower_change'].shape == (n_combos, len(data))
    assert len(result['jaws_lengths']) == n_combos

def test_gatorosc_consistency():
    """Test that GatorOsc produces consistent results."""
    np.random.seed(42)
    data = np.random.randn(100) * 10 + 50
    
    # Calculate twice with same parameters
    upper1, lower1, upper_change1, lower_change1 = ta.gatorosc(data)
    upper2, lower2, upper_change2, lower_change2 = ta.gatorosc(data)
    
    # Results should be identical
    np.testing.assert_array_equal(upper1, upper2)
    np.testing.assert_array_equal(lower1, lower2)
    np.testing.assert_array_equal(upper_change1, upper_change2)
    np.testing.assert_array_equal(lower_change1, lower_change2)

def test_gatorosc_kernel_options():
    """Test GatorOsc with different kernel options."""
    data = np.random.randn(100) * 10 + 50
    
    # Test with different kernels
    for kernel in [None, "scalar", "avx2", "avx512"]:
        try:
            upper, lower, upper_change, lower_change = ta.gatorosc(data, kernel=kernel)
            assert len(upper) == len(data)
            assert len(lower) == len(data)
        except ValueError:
            # Some kernels might not be supported on this platform
            pass

def test_gatorosc_nan_handling():
    """Test GatorOsc handling of NaN values in input."""
    data = np.random.randn(100) * 10 + 50
    data[40:45] = np.nan  # Insert some NaN values
    
    # Should still work
    upper, lower, upper_change, lower_change = ta.gatorosc(data)
    
    assert len(upper) == len(data)
    assert len(lower) == len(data)
    assert len(upper_change) == len(data)
    assert len(lower_change) == len(data)
    
    # Should have some valid values despite NaNs in input
    assert not np.all(np.isnan(upper))
    assert not np.all(np.isnan(lower))

def test_gatorosc_value_ranges():
    """Test that GatorOsc outputs are in expected ranges."""
    # Create trending data
    trend_up = np.linspace(40, 60, 100) + np.random.randn(100) * 0.5
    trend_down = np.linspace(60, 40, 100) + np.random.randn(100) * 0.5
    
    # Test with uptrend
    upper_up, lower_up, _, _ = ta.gatorosc(trend_up)
    
    # Test with downtrend
    upper_down, lower_down, _, _ = ta.gatorosc(trend_down)
    
    # Both should have valid values
    assert not np.all(np.isnan(upper_up))
    assert not np.all(np.isnan(lower_up))
    assert not np.all(np.isnan(upper_down))
    assert not np.all(np.isnan(lower_down))

File: tests/wasm/test_gatorosc.js
/**
 * WASM binding tests for GatorOsc indicator.
 * These tests ensure WASM bindings work correctly for the 4-output Gator Oscillator.
 */
import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import { 
    loadTestData, 
    assertArrayClose, 
    assertClose,
    isNaN,
    assertAllNaN,
    assertNoNaN
} from './test_utils.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let wasm;
let testData;

test.before(async () => {
    
    try {
        const wasmPath = path.join(__dirname, '../../pkg/vector_ta.js');
        const importPath = process.platform === 'win32' 
            ? 'file:///' + wasmPath.replace(/\\/g, '/')
            : wasmPath;
        wasm = await import(importPath);
    } catch (error) {
        console.error('Failed to load WASM module. Run "wasm-pack build --features wasm --target nodejs" first');
        throw error;
    }
    
    testData = loadTestData();
});

test('GatorOsc basic functionality', () => {
    const close = new Float64Array(testData.close);
    
    
    const result = wasm.gatorosc_js(close, 13, 8, 8, 5, 5, 3);
    
    
    assert.strictEqual(result.rows, 4, 'Should have 4 output rows');
    assert.strictEqual(result.cols, close.length, 'Should have same length as input');
    assert.strictEqual(result.values.length, 4 * close.length, 'Should have 4x input length values');
    
    
    const upper = result.values.slice(0, close.length);
    const lower = result.values.slice(close.length, 2 * close.length);
    const upperChange = result.values.slice(2 * close.length, 3 * close.length);
    const lowerChange = result.values.slice(3 * close.length);
    
    
    let hasValidUpper = false;
    let hasValidLower = false;
    
    for (let i = 20; i < upper.length; i++) {
        if (!isNaN(upper[i])) {
            assert(upper[i] >= 0, `Upper value at ${i} should be non-negative: ${upper[i]}`);
            hasValidUpper = true;
        }
        if (!isNaN(lower[i])) {
            assert(lower[i] <= 0, `Lower value at ${i} should be non-positive: ${lower[i]}`);
            hasValidLower = true;
        }
    }
    
    assert(hasValidUpper, 'Should have valid upper values after warmup');
    assert(hasValidLower, 'Should have valid lower values after warmup');
});

test('GatorOsc fast API (no aliasing)', () => {
    const data = new Float64Array([50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60]);
    const len = data.length;
    
    
    const inPtr = wasm.gatorosc_alloc(len);
    const upperPtr = wasm.gatorosc_alloc(len);
    const lowerPtr = wasm.gatorosc_alloc(len);
    const upperChangePtr = wasm.gatorosc_alloc(len);
    const lowerChangePtr = wasm.gatorosc_alloc(len);
    
    try {
        
        const memory = new Float64Array(wasm.__wasm.memory.buffer);
        memory.set(data, inPtr / 8);
        
        
        wasm.gatorosc_into(
            inPtr, upperPtr, lowerPtr, upperChangePtr, lowerChangePtr,
            len, 5, 3, 3, 2, 2, 1
        );
        
        
        const memory2 = new Float64Array(wasm.__wasm.memory.buffer);
        
        
        const upper = Array.from(memory2.slice(upperPtr / 8, upperPtr / 8 + len));
        const lower = Array.from(memory2.slice(lowerPtr / 8, lowerPtr / 8 + len));
        
        
        assert(upper.some(v => !isNaN(v)), 'Upper should have some non-NaN values');
        assert(lower.some(v => !isNaN(v)), 'Lower should have some non-NaN values');
        
    } finally {
        
        wasm.gatorosc_free(inPtr, len);
        wasm.gatorosc_free(upperPtr, len);
        wasm.gatorosc_free(lowerPtr, len);
        wasm.gatorosc_free(upperChangePtr, len);
        wasm.gatorosc_free(lowerChangePtr, len);
    }
});

test('GatorOsc fast API (with aliasing)', () => {
    const data = new Float64Array([50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60]);
    const len = data.length;
    
    
    const inPtr = wasm.gatorosc_alloc(len);
    const lowerPtr = wasm.gatorosc_alloc(len);
    const upperChangePtr = wasm.gatorosc_alloc(len);
    const lowerChangePtr = wasm.gatorosc_alloc(len);
    
    try {
        
        const memory = new Float64Array(wasm.__wasm.memory.buffer);
        memory.set(data, inPtr / 8);
        
        
        wasm.gatorosc_into(
            inPtr, inPtr, lowerPtr, upperChangePtr, lowerChangePtr,  
            len, 5, 3, 3, 2, 2, 1
        );
        
        
        const memory2 = new Float64Array(wasm.__wasm.memory.buffer);
        
        
        const upper = Array.from(memory2.slice(inPtr / 8, inPtr / 8 + len));  
        const lower = Array.from(memory2.slice(lowerPtr / 8, lowerPtr / 8 + len));
        
        
        assert(upper.some(v => !isNaN(v)), 'Upper should have some non-NaN values');
        assert(lower.some(v => !isNaN(v)), 'Lower should have some non-NaN values');
        
    } finally {
        
        wasm.gatorosc_free(inPtr, len);
        wasm.gatorosc_free(lowerPtr, len);
        wasm.gatorosc_free(upperChangePtr, len);
        wasm.gatorosc_free(lowerChangePtr, len);
    }
});

test('GatorOsc error handling', () => {
    
    assert.throws(() => {
        wasm.gatorosc_js(new Float64Array([]), 13, 8, 8, 5, 5, 3);
    }, /empty|invalid|insufficient/i);
    
    
    const nanData = new Float64Array(50).fill(NaN);
    assert.throws(() => {
        wasm.gatorosc_js(nanData, 13, 8, 8, 5, 5, 3);
    }, /nan|invalid/i);
    
    
    const data = new Float64Array([1, 2, 3, 4, 5]);
    assert.throws(() => {
        wasm.gatorosc_js(data, 0, 8, 8, 5, 5, 3);  
    }, /invalid|parameter|length/i);
});

test('GatorOsc batch processing', () => {
    const close = new Float64Array(testData.close.slice(0, 100));  
    
    const config = {
        jaws_length_range: [10, 15, 5],
        jaws_shift_range: [6, 10, 2],
        teeth_length_range: [6, 10, 2],
        teeth_shift_range: [3, 6, 3],
        lips_length_range: [3, 6, 3],
        lips_shift_range: [2, 4, 2]
    };
    
    const result = wasm.gatorosc_batch(close, config);
    
    
    assert(result.rows > 0, 'Should have parameter combinations');
    assert.strictEqual(result.cols, close.length, 'Should match input length');
    assert.strictEqual(result.outputs, 4, 'Should have 4 outputs');
    assert.strictEqual(result.values.length, 4 * result.rows * result.cols, 'Values array size should match');
    assert(Array.isArray(result.combos), 'Should have combos array');
    assert.strictEqual(result.combos.length, result.rows, 'Combos should match rows');
    
    
    for (const combo of result.combos) {
        assert(combo.jaws_length !== undefined, 'Combo should have jaws_length');
        assert(combo.jaws_shift !== undefined, 'Combo should have jaws_shift');
        assert(combo.teeth_length !== undefined, 'Combo should have teeth_length');
        assert(combo.teeth_shift !== undefined, 'Combo should have teeth_shift');
        assert(combo.lips_length !== undefined, 'Combo should have lips_length');
        assert(combo.lips_shift !== undefined, 'Combo should have lips_shift');
    }
});

test('GatorOsc consistency', () => {
    const data = new Float64Array([
        50.0, 51.0, 52.0, 51.5, 53.0, 54.0, 53.5, 55.0, 56.0, 55.5,
        57.0, 58.0, 57.5, 59.0, 60.0, 59.5, 58.0, 57.0, 58.5, 59.0
    ]);
    
    
    const result1 = wasm.gatorosc_js(data, 5, 3, 3, 2, 2, 1);
    const result2 = wasm.gatorosc_js(data, 5, 3, 3, 2, 2, 1);
    
    
    assertArrayClose(
        result1.values, 
        result2.values, 
        1e-10,
        "GatorOsc should produce consistent results"
    );
});

test('GatorOsc memory allocation/deallocation', () => {
    
    for (let i = 0; i < 10; i++) {
        const len = 100 + i * 10;
        const ptr = wasm.gatorosc_alloc(len);
        assert(ptr > 0, 'Should return valid pointer');
        wasm.gatorosc_free(ptr, len);
    }
    
    
    wasm.gatorosc_free(0, 100);  
});